src/extension.ts

```diff
@@ -519,6 +519,72 @@ function launchMetals(
       });
     });
 
+    registerCommand("metals.new-scala-file", (dirArg: Uri) => {
+      const dir = (() => {
+        if (dirArg !== undefined) {
+          return dirArg.toString();
+        } else if (window.activeTextEditor) {
+          const currentPath = window.activeTextEditor.document.uri.fsPath;
+          return Uri.file(path.dirname(currentPath)).toString();
+        }
+        // NOTE(aleksei): I would return workspace folder here, but not sure if there's API for that.
+        // Moving this funcitonality to server.
+        return undefined;
+      })();
+
+      const worksheetPick = { label: 'Worksheet' }
+      const classPick = { label: 'Class', kind: 'class' }
+      const objectPick = { label: 'Object', kind: 'object' }
+      const traitPick = { label: 'Trait', kind: 'trait' }
+
+      function withName(f: (name: string) => any, placeHolder?: string) {
+        return window.showInputBox({ placeHolder: placeHolder }).then(name => {
+          if (name !== undefined)
+            f(name)
+        });
+      }
+
+      return window.showQuickPick(
+        [classPick, worksheetPick],
+        { placeHolder: 'Select the kind of file to create' }
+      ).then(fileKindPick => {
+        switch (fileKindPick) {
+          case worksheetPick:
+            return withName(name => {
+              client.sendRequest(ExecuteCommandRequest.type, {
+                command: "new-scala-worksheet",
+                arguments: [dir, name]
+              }).then(result => {
+                workspace
+                  .openTextDocument(Uri.parse(result))
+                  .then(textDocument => window.showTextDocument(textDocument));
+              });
+            });
+          case classPick:
+            return window.showQuickPick(
+              //TODO: maybe add descriptions?
+              [classPick, objectPick, traitPick]
+            ).then(classKindPick => {
+              switch (classKindPick) {
+                case classPick:
+                case objectPick:
+                case traitPick:
+                  return withName(name => {
+                    client.sendRequest(ExecuteCommandRequest.type, {
+                      command: "new-scala-class",
+                      arguments: [dir, name, classKindPick.kind]
```

Comment:
We should be able to have an object as an argument here and we can reduce it all to a single pick. Worksheet/object/trait/class together.

Reply:
Done